Fix issues in cloud prepare for Azure and RHOS

aswinsuryan opened the pull request:
Fix issues in the cloud prepare for Azure and RHOS

1) Fix Azure machine set name now changed to region + uuid
and limited to 20 characters to prevent issues long cluster
name

2) Machine set creation and deletion logic updated. Now it lists
machines set to check for existing dedicated nodes and clean up
gateway nodes along with tagged g/w nodes. Earlier it listed
only tagged nodes. This resulted in the creation of multiple g/w nodes
cases when cloud prepare is run multiple times due to a delay in the creation
g/w nodes from the machine set.

Fixes: https://github.com/stolostron/backlog/issues/25111

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our developer guide: https://submariner.io/development/
2. Ensure you have added the appropriate tests for your PR: https://submariner.io/development/code-review/#test-new-functionality
3. Read the code review guide to ease the review process: https://submariner.io/development/code-review/
4. If the PR is unfinished, mark it as a draft: https://submariner.io/development/code-review/#mark-work-in-progress-prs-as-drafts
5. If you are using CI to debug, use your private fork: https://submariner.io/development/code-review/#use-private-forks-for-debugging-prs-by-running-ci
6. Add labels to the PR as appropriate.

This template is based on the K8s/K8s template:

https://github.com/kubernetes/kubernetes/blob/master/.github/PULL_REQUEST_TEMPLATE.md
-->

tpantelis commented on pkg/azure/ocpgwdeployer.go:
```diff
@@ -267,7 +267,7 @@ func (d *ocpGatewayDeployer) initMachineSet(name, zone string) (*unstructured.Un
 
 func (d *ocpGatewayDeployer) deployGateway(zone string) error {
-	machineSet, err := d.initMachineSet(MachineName(d.azure.InfraID, d.azure.Region, zone), zone)
+	machineSet, err := d.initMachineSet(MachineName(d.azure.Region, zone), zone)
```
If we just need any valid, unique name for the machine set, why not just use a UUID or use the K8s generate name functionality?

aswinsuryan replied:
the public IP will have this cluster name added to it. So the UUID should also be less than 20 characters[1].
Also when listing machineset wouldn't it be better to have a submariner identifier in it?

[1] https://github.com/openshift/machine-api-provider-azure/blob/main/pkg/cloud/azure/defaults.go#L86

skitt commented on pkg/azure/ocpgwdeployer.go:
```diff
@@ -7,7 +7,7 @@ Licensed under the Apache License, Version 2.0 (the "License");
 You may obtain a copy of the License at
 
-http://www.apache.org/licenses/LICENSE-2.0
+    http://www.apache.org/licenses/LICENSE-2.0
```
Please don’t change this.

aswinsuryan replied:
I will revert,  I was just testing the PR. I think I did not raise the PR from the latest code.

skitt commented on pkg/azure/ocpgwdeployer.go:
```diff
@@ -276,15 +276,13 @@ func (d *ocpGatewayDeployer) deployGateway(zone string) error {
+// We add "-subgw-", 7 characters, for a total of 20 with the hyphen between region and zone.
+func MachineName(region, zone string) string {
+	if len(region)+len(zone) > 12 {
 		// Limit the name length to 40 characters
```
This comment needs to be updated too.

aswinsuryan replied:
Done.

skitt commented on pkg/azure/ocpgwdeployer.go:
```diff
@@ -318,7 +312,7 @@ func (d *ocpGatewayDeployer) getAvailabilityZones(gwNodes []v1.Node) (stringset.
 
 	pager := resourceSKUClient.NewListPager(&armcompute.ResourceSKUsClientListOptions{
-		Filter: pointer.String(d.azure.Region),
+		Filter: to.StringPtr(d.azure.Region),
```
go-autorest is deprecated and will no longer be supported soon, please don’t change this.

aswinsuryan replied:
Done. Sorry, it was not intentional.

sridhargaddam commented on pkg/azure/ocpgwdeployer.go:
```diff
@@ -276,30 +276,24 @@ func (d *ocpGatewayDeployer) deployGateway(zone string) error {
 	}
 
-	return infraID + submarinerGatewayGW + region + "-" + zone
+	return submarinerGatewayGW + region + "-" + zone
```
I think this would create issues during upgrades. 

Say, we initially had an existing deployment where the Gatewaynode is created using a MachineSet. Following this PR if we change the MachineSet name, the reconciler code (if triggered from ACM or if user manually executes subctl cloud-prepare once again), will look for a MachineSet with a specific name and when it does not find it, will re-create it. Can you please explain how we handle this scenario?

aswinsuryan replied:
We list the g/w nodes [1] using the submariner g/w label and if we have a sufficient number of nodes we skip the creation. For deletion we use the same API and check for the substring subgw- (earlier it was -subgw-) and it should work as well, 

[1]https://github.com/submariner-io/cloud-prepare/blob/devel/pkg/azure/ocpgwdeployer.go#L99

sridhargaddam replied:
I see that we are not using this API or the submarinerGatewayGW as part of delete*, but identify the machineSet/prefix/name using the "-" char. IMHO this is fragile. Can you please fix it or add a comment in the cleanup path about this dependency.

* https://github.com/submariner-io/cloud-prepare/blob/4b2f851014e5b98b47cd7c69e36dcc4e62b245f8/pkg/azure/ocpgwdeployer.go#L403

aswinsuryan replied:
I will modify the delete code and fill include the fix for multiple g/w nodes getting spawned in Azure cluster.

aswinsuryan replied:
Updated the delete logic.

sridhargaddam commented on pkg/azure/ocpgwdeployer.go:
```diff
@@ -276,30 +276,24 @@ func (d *ocpGatewayDeployer) deployGateway(zone string) error {
-			zone = zone[0:6]
+// We add "subgw-", 6 characters, for a total of 20 with the hyphen between region and zone.
+func MachineName(region, zone string) string {
+	if len(region)+len(zone) > 12 {
```
If we are using the region/zone + some submariner identifier in the names, it means we can only create a limited number of Gateway nodes and cannot scale them beyond the number of zones. I think this is not specific to this PR/Azure but is applicable to all the other cloud-prepare methods.

Ideally, I feel we should use some UUID for the names (with a suitable length) and label (or tag) the MachineSetObjects to identify the Submariner resources rather than creating names with some pattern which is error prone.

aswinsuryan replied:
we decided to use Zone assuming we will have one g/w per availability zone. We follow this for all cloud expect Openstack. Can we raise a separate issue and track it over there?

sridhargaddam replied:
Yes, please report an issue for this. Thanks.

aswinsuryan replied:
https://github.com/submariner-io/cloud-prepare/issues/532

maayanf24 commented:
@aswinsuryan - please add release notes

tpantelis commented on pkg/azure/ocpgwdeployer.go:
```diff
@@ -360,7 +349,7 @@ func (d *ocpGatewayDeployer) Cleanup(status reporter.Interface) error {
-	err = d.deleteGateway()
+	err = d.deleteGateway(status)
 	if err != nil {
 		return status.Error(err, "removing gateway node failed")
```
`deleteGateway` now logs errors to the `status` so we should just return the error here.

tpantelis commented on pkg/azure/ocpgwdeployer.go:
```diff
@@ -386,32 +373,81 @@ func (d *ocpGatewayDeployer) deleteGateway() error {
-				return errors.Wrapf(err, "failed to remove labels from worker node")
-			}
+	for i := range machineSetList {
+		status.Start(fmt.Sprintf("Deleting the gateway instance %q", machineSetList[i].GetName()))
```
```suggestion
		status.Start("Deleting the gateway instance %q", machineSetList[i].GetName())
```

tpantelis commented on pkg/azure/ocpgwdeployer.go:
```diff
@@ -386,32 +373,81 @@ func (d *ocpGatewayDeployer) deleteGateway() error {
+
+		err = d.deletePublicIP(ctx, pubIPClient, publicIPName)
+		if err != nil {
+			return errors.Wrapf(err, "failed to delete public-ip")
```
```suggestion
			return status.Error(err, "failed to delete public-ip %q", publicIPName)
```

tpantelis commented on pkg/azure/ocpgwdeployer.go:
```diff
@@ -386,32 +373,81 @@ func (d *ocpGatewayDeployer) deleteGateway() error {
-				return err
-			}
+		if err != nil {
+			return errors.Wrapf(err, "failed to cleanup node %q", gwNodes[i].Name)
```
We should be consistent returning `status.Error` vs `errors.Wrapf.`

```suggestion
			return status.Error(err, "failed to cleanup node %q", gwNodes[i].Name)
```

tpantelis commented on pkg/azure/ocpgwdeployer.go:
```diff
@@ -386,32 +373,81 @@ func (d *ocpGatewayDeployer) deleteGateway() error {
-			return errors.Wrapf(d.msDeployer.Delete(machineSet), "error deleting machine set %q", machineSet.GetName())
+		err = d.deletePublicIP(ctx, pubIPClient, publicIPName)
+		if err != nil {
+			return errors.Wrapf(err, "failed to delete public-ip")
```
```suggestion
			return status.Error(err, "failed to delete public-ip")
```

tpantelis commented on pkg/azure/ocpgwdeployer.go:
```diff
@@ -79,40 +79,36 @@ func (d *ocpGatewayDeployer) Deploy(input api.GatewayDeployInput, status reporte
+	nsgClient, nwClient, pubIPClient, err := d.getClients(status)
 	if err != nil {
-		return status.Error(err, "Failed to get network security groups client")
+		return status.Error(err, "Failed to get client")
```
`getClients` logs errors.

```suggestion
		return err
```

tpantelis commented on pkg/azure/ocpgwdeployer.go:
```diff
@@ -79,40 +79,36 @@ func (d *ocpGatewayDeployer) Deploy(input api.GatewayDeployInput, status reporte
+	machineSets, err := d.msDeployer.List()
 	if err != nil {
-		return status.Error(err, "Failed to get network public IP addresses client")
+		return errors.Wrap(err, "error getting the gateway machinesets")
```
For consistency...

```suggestion
		return status.Error(err, "error getting the gateway machinesets")
```

tpantelis commented on pkg/k8s/k8siface.go:
```diff
@@ -126,3 +128,15 @@ func (k *k8sIface) RemoveGWLabelFromWorkerNode(node *v1.Node) error {
+	var resultNode []v1.Node
+
+	for i := 0; i < len(gwNodes); i++ {
+		if !(gwNodes[i].GetLabels()[ocp.MachineSetOwnerLabel] == "submariner") {
```
Easier to read...

```suggestion
		if gwNodes[i].GetLabels()[ocp.MachineSetOwnerLabel] != "submariner" {
```

tpantelis commented on pkg/k8s/k8siface.go:
```diff
@@ -126,3 +128,15 @@ func (k *k8sIface) RemoveGWLabelFromWorkerNode(node *v1.Node) error {
 	})
 }
+
+func (k *k8sIface) FilterDedicatedNode(gwNodes []v1.Node) []v1.Node {
```
This API is a wrapper for K8s so this function doesn't seem to belong here. It is tied to machine sets  and references the `ocp` package and so it seems better suited there, ie `ocp.FilterDedicatedNodes`.

tpantelis commented on pkg/ocp/machinesets_test.go:
```diff
@@ -193,22 +193,34 @@ var _ = Describe("K8s MachineSetDeployer", func() {
+				"spec": map[string]interface{}{
+					"metadata": map[string]interface{}{
+						"labels": map[string]interface{}{
+							"submariner.io/machine-set-owner": owner,
```
Use constant here.

tpantelis commented on pkg/rhos/ocpgwdeployer.go:
```diff
@@ -156,11 +155,18 @@ func (d *ocpGatewayDeployer) Deploy(input api.GatewayDeployInput, status reporte
 
+	machineSets, err := d.msDeployer.List()
+	if err != nil {
+		return errors.Wrap(err, "error getting the gateway machinesets")
```
```suggestion
		return status.Error(err, "error getting the gateway machinesets")
```

tpantelis commented on pkg/rhos/ocpgwdeployer.go:
```diff
@@ -156,11 +155,18 @@ func (d *ocpGatewayDeployer) Deploy(input api.GatewayDeployInput, status reporte
 	gwNodes, err := d.K8sClient.ListGatewayNodes()
 	if err != nil {
-		return status.Error(err, "listing the existing gateway nodes failed")
+		return errors.Wrap(err, "listing the existing gateway nodes failed")
```
```suggestion
		return status.Error(err, "listing the existing gateway nodes failed")
```

tpantelis commented on pkg/rhos/ocpgwdeployer.go:
```diff
@@ -316,11 +316,7 @@ func (d *ocpGatewayDeployer) Cleanup(status reporter.Interface) error {
-		// the gateway node was deployed using the OCPMachineSet which is already handled above and
-		// the node should be already be in deleting state.
-		if !strings.HasPrefix(gwNodes[i].Name, d.InfraID) || strings.HasPrefix(gwNodes[i].Name, prefix) {
+		if gwNodes[i].GetLabels()[ocp.MachineSetOwnerLabel] == "submariner" {
```
Should make "submariner" a constant in `ocp` as it's referenced in several places, eg `LabelValueOwnedBy`.

aswinsuryan commented:
@sridhargaddam @tpantelis  Each vm is created(will have same name as machine set) in its own resource group. So having a same name vm in different clusters should not be an issue. Should we still change vm name to just uuid?

sridhargaddam commented on pkg/azure/gw-machineset.go:
```diff
@@ -44,6 +44,7 @@ spec:
       metadata:
         labels:
           submariner.io/gateway: "true"
+          submariner.io/machine-set-owner: "submariner"
```
As discussed on the call, lets get rid of this new label as its introducing lot of corner cases.

aswinsuryan replied:
Done.

sridhargaddam reviewed (changes requested):
Thanks for the PR @aswinsuryan

sridhargaddam commented on pkg/ocp/machinesets.go:
```diff
@@ -207,19 +206,28 @@ func (msd *k8sMachineSetDeployer) List() ([]unstructured.Unstructured, error) {
+		addToResult := true
+
+		for i := 0; i < len(machineSets); i++ {
+			if strings.Contains(gwNodes[i].GetName(), machineSets[i].GetName()) {
```
Can we safely assume that machineSet.name will be same as gwNode.Name?

Following is the output from one of the deployments.

```
[sgaddam@localhost submariner]$ oc get nodes
NAME         STATUS   ROLES                  AGE   VERSION
master-0-0   Ready    control-plane,master   19d   v1.25.4+77bec7a
master-0-1   Ready    control-plane,master   19d   v1.25.4+77bec7a
master-0-2   Ready    control-plane,master   19d   v1.25.4+77bec7a
worker-0-0   Ready    worker                 19d   v1.25.4+77bec7a
worker-0-1   Ready    worker                 19d   v1.25.4+77bec7a

```

```
[sgaddam@localhost submariner]$ oc get machinesets -A
NAMESPACE               NAME                     DESIRED   CURRENT   READY   AVAILABLE   AGE
openshift-machine-api   ice01-0-qjgw7-worker-0   2         2         2       2           19d

```

aswinsuryan replied:
I think it should depend on the machineset provider, for Azure and RHOS that is the pattern I observed. We may need an extra label, if we are planning not using the name.

sridhargaddam replied:
Even if its a pattern today, things might change anytime. So, IMHO we should identify alternatives to this.

Please check if there is a way to know the node/machine-name from the machineset - I think there should be one (the machineset controller should be using it to identify the associated machines).

aswinsuryan replied:
Just checked, 

the below one to get info about node from machineset return emty

`% kubectl get machinesets -n openshift-machine-api subgw-eastus-05ea1c -o yaml | grep subgw-eastus-05ea1c-6475l`

and when tried the otherway , it does not have any reference to the machine name.

```
% kubectl get nodes subgw-eastus-05ea1c-6475l -o yaml | grep  subgw-eastus-05ea1c                                         
    cloud.network.openshift.io/egress-ipconfig: '[{"interface":"subgw-eastus-05ea1c-6475l-nic","ifaddr":{"ipv4":"10.0.128.0/17"},"capacity":{"ip":255}}]'
    csi.volume.kubernetes.io/nodeid: '{"disk.csi.azure.com":"subgw-eastus-05ea1c-6475l","file.csi.azure.com":"subgw-eastus-05ea1c-6475l"}'
    machine.openshift.io/machine: openshift-machine-api/subgw-eastus-05ea1c-6475l
    kubernetes.io/hostname: subgw-eastus-05ea1c-6475l
  name: subgw-eastus-05ea1c-6475l
  providerID: azure:///subscriptions/3c7a4fa5-8496-4cdc-8a78-e344e2f896d7/resourceGroups/asuryana1-tbrrj-rg/providers/Microsoft.Compute/virtualMachines/subgw-eastus-05ea1c-6475l
  - address: subgw-eastus-05ea1c-6475l
```

sridhargaddam replied:
Okay, I see that we have an annotation on the nodes as shown below. But this may not be any different to the assumption we are making about the names. 

```
machine.openshift.io/machine: openshift-machine-api/ice01-1-pgnls-master-0
```

sridhargaddam commented on pkg/azure/ocpgwdeployer.go:
```diff
@@ -96,7 +95,7 @@ func (d *ocpGatewayDeployer) Deploy(input api.GatewayDeployInput, status reporte
 
 	gwNodeItems := gwNodes.Items
-	taggedExistingNodes := ocp.FilterDedicatedNode(gwNodeItems)
+	taggedExistingNodes := ocp.RemoveDuplicates(nil, gwNodeItems)
```
why nil?

aswinsuryan replied:
It was an error while refactoring. Fixed it.

sridhargaddam commented on pkg/ocp/machinesets_test.go:
```diff
@@ -176,7 +176,7 @@ var _ = Describe("K8s MachineSetDeployer", func() {
 				_, err := msClient.Create(context.TODO(), machineSet, metav1.CreateOptions{})
 				Expect(err).To(Succeed())
-				machineSet = newMachineSet("other")
+				machineSet = newMachineSet("fasle")
```
Is this typo or intentional?

aswinsuryan replied:
no .. :D

aswinsuryan replied:
Done

aswinsuryan committed:
Fix issues in cloud prepare for Azure and RHOS

1) Fix Azure machine set name now changed to region + uuid
and limited to 20 characters to prevent issues long cluster
name

2) Machine set creation and deletion logic updated. Now it lists
machines set to check for existing dedicated nodes and cleaning up
gateway nodesi along with tagged g/w nodes. Earlier it listed
only tagged nodes . This resulted in creation of  multiple g/w nodes
cases when cloud prepare is run multiple times due to delay in creation
g/w nodes from the machine set.

Signed-off-by: Aswin Suryanarayanan <aswinsuryan@gmail.com>

sridhargaddam commented:
@aswinsuryan please update the commit message in the PR.

aswinsuryan commented:
> @aswinsuryan please update the commit message in the PR.

@sridhargaddam done.
